File: cpp/08_063.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    // Find the kth element in two sorted arrays
    int findKthElement(vector<int>& nums1, vector<int>& nums2, 
                      int k) {
        // Ensure nums1 is the smaller array
        if (nums1.size() > nums2.size()) {
            return findKthElement(nums2, nums1, k);
        }
        
        int m = nums1.size();
        int n = nums2.size();
        // Set binary search bounds
        int left = max(0, k - n);
        int right = min(k, m);
        
        // Binary search
        while (left <= right) {
            int mid1 = (left + right) / 2;
            int mid2 = k - mid1;
            
            // Get boundary elements
            int l1 = (mid1 == 0) ? INT_MIN : nums1[mid1 - 1];
            int r1 = (mid1 == m) ? INT_MAX : nums1[mid1];
            int l2 = (mid2 == 0) ? INT_MIN : nums2[mid2 - 1];
            int r2 = (mid2 == n) ? INT_MAX : nums2[mid2];
            
            // Check if partition is correct
            if (l1 <= r2 && l2 <= r1) {
                return max(l1, l2);
            } 
            // Adjust search boundaries
            else if (l1 > r2) {
                right = mid1 - 1;
            } 
            else {
                left = mid1 + 1;
            }
        }
        
        // Should not reach here if arrays are sorted
        return -1;
    }
};

File: python/08_063.py
def findKthElement(nums1, nums2, k):

    # Ensure nums1 is the smaller array for simpler binary search
    if len(nums1) > len(nums2):
        return findKthElement(nums2, nums1, k)
    
    m, n = len(nums1), len(nums2)
    left, right = max(0, k - n), min(k, m)
    
    while left <= right:
        mid1 = (left + right) // 2
        mid2 = k - mid1
        
        l1 = float('-inf') if mid1 == 0 else nums1[mid1 - 1]
        r1 = float('inf') if mid1 == m else nums1[mid1]
        l2 = float('-inf') if mid2 == 0 else nums2[mid2 - 1]
        r2 = float('inf') if mid2 == n else nums2[mid2]
        
        if l1 <= r2 and l2 <= r1:
            return max(l1, l2)
        elif l1 > r2:
            right = mid1 - 1
        else:
            left = mid1 + 1
    
    return -1  # Should not reach here if arrays are sorted